sdr_airspy.c

```diff
@@ -0,0 +1,467 @@
+    } else {
+        return false;
+    }
+}
+
+void airspyShowHelp()
+{
+    printf("      airspy-specific options (use with --device-type airspy)\n");
+    printf("\n");
+    printf("--serial-number <0x...>  Serial number (in hex) of Airspy to open (optional)\n");
+    printf("--enable-lna-agc         Enable LNA AGC (default off)\n");
+    printf("--enable-mixer-agc       Enable MIXER AGC (default off)\n");
+    printf("--linearity-gain <n>     Set linearity simplified gain (0-21)\n");
+    printf("--sensitivity-gain <n>   Set sensitivity simplified gain (0-21)\n");
+    printf("--mixer-gain <n>         Set MIXER gain (0-15, default 8)\n");
+    printf("--lna-gain <n>           Set LNA gain (0-14, default 13)\n");
+    printf("--vga-gain <n>           Set VGA gain (0-15, default 5)\n");
+    printf("--enable-airspy-biast    Set bias tee supply on (default off)\n");
+    printf("\n");
+}
+
+bool airspyHandleOption(int argc, char **argv, int *jptr)
+{
+    int j = *jptr;
+    bool result = true;
+    bool more = (j + 1 < argc);
+
+    if (!strcmp(argv[j], "--serial-number") && more) {
+        result = parse_uintt(argv[++j], &AIRSPY.serial_number, 64);
+    } else if (!strcmp(argv[j], "--enable-lna-agc")) {
+        AIRSPY.enable_lna_agc = true;
+    } else if (!strcmp(argv[j], "--enable-mixer-agc")) {
+        AIRSPY.enable_mixer_agc = true;
+    } else if (!strcmp(argv[j], "--enable-airspy-biast")) {
+        AIRSPY.enable_airspy_biast = true;
+    } else if (!strcmp(argv[j], "--linearity-gain") && more) {
+        result = parse_uintt(argv[++j], (uint64_t *)&AIRSPY.linearity_gain, 8);
+        AIRSPY.enable_linearity = true;
+    } else if (!strcmp(argv[j], "--sensitivity-gain") && more) {
+        result = parse_uintt(argv[++j], (uint64_t *)&AIRSPY.sensitivity_gain, 8);
+        AIRSPY.enable_sensitivity = true;
+    } else if (!strcmp(argv[j], "--mixer-gain") && more) {
+        result = parse_uintt(argv[++j], (uint64_t *)&AIRSPY.mixer_gain, 8);
+    } else if (!strcmp(argv[j], "--lna-gain") && more) {
+        result = parse_uintt(argv[++j], (uint64_t *)&AIRSPY.lna_gain, 8);
+    } else if (!strcmp(argv[j], "--vga-gain") && more) {
+        result = parse_uintt(argv[++j], (uint64_t *)&AIRSPY.vga_gain, 8);
+    } else {
+        return false;
+    }
+
+    if (!result)
+        return false;
+
+    if (AIRSPY.enable_linearity && AIRSPY.enable_sensitivity) {
+        fprintf(stderr,
+            "Airspy config: --linearity-gain and --sensitivity-gain cannot be used together.\n");
+        return false;
+    }
+
+    if ((AIRSPY.enable_linearity || AIRSPY.enable_sensitivity) && (AIRSPY.enable_lna_agc || AIRSPY.enable_mixer_agc)) {
+        fprintf(stderr,
+            "Airspy config: --linearity-gain/--sensitivity-gain cannot be used with --enable-lna-agc/--enable-mixer-agc.\n");
+        return false;
+    }
+
+    *jptr = j;
+    return true;
+}
+
+bool airspyOpen(void) {
+    #define AIRSPY_STATUS(status, message) \
```

Comment:
You probably want an error-exit label that does teardown rather than just returning; there is a bunch of other teardown that might be needed depending on how far you get (freeing the soxr stuff and your buffers etc)